polyfill/lib/duration.mjs

```diff
@@ -279,69 +282,124 @@ export class Duration {
     const maximum = maximumIncrements[smallestUnit];
     if (maximum !== undefined) ES.ValidateTemporalRoundingIncrement(roundingIncrement, maximum, false);
 
+    let zonedRelativeTo = ES.IsTemporalZonedDateTime(relativeTo) ? relativeTo : undefined;
+    let plainRelativeTo = ES.IsTemporalDate(relativeTo) ? relativeTo : undefined;
+
+    const roundingGranularityIsNoop = smallestUnit === 'nanosecond' && roundingIncrement === 1;
+    const balancingRequested = largestUnit !== existingLargestUnit;
+    const calendarUnitsPresent = years !== 0 || months !== 0 || weeks !== 0;
+    const timeUnitsOverflowWillOccur =
+      minutes >= 60 || seconds >= 60 || milliseconds >= 1000 || microseconds >= 1000 || nanoseconds >= 1000;
+    const hoursToDaysConversionMayOccur = (days !== 0 && zonedRelativeTo) || hours >= 24;
+    if (
+      roundingGranularityIsNoop &&
+      !balancingRequested &&
+      !calendarUnitsPresent &&
+      !timeUnitsOverflowWillOccur &&
+      !hoursToDaysConversionMayOccur
+    ) {
+      return new Duration(years, months, weeks, days, hours, minutes, seconds, milliseconds, microseconds, nanoseconds);
+    }
+
+    const plainRelativeToWillBeUsed =
+      smallestUnit === 'year' ||
+      smallestUnit === 'month' ||
+      smallestUnit === 'week' ||
```

Comment:
This pattern shows up a few times, would it be worth making a function like:

`unitIsYearMonthWeek = (unit) => unit === 'year' || unit === 'month' || unit === 'week'`

Reply:
Good point, I think we can add an abstract operation for this in the spec as well. It will make the lines less long, and easier to read. I'd prefer to do this as an editorial update after this normative PR lands, though. -> #2692